### PR TITLE
prov/efa: Handling CUDA buffers in send/recv paths (new)

### PR DESCRIPTION
**Pull request by wzamazon**
This is the update of 

https://github.com/ofiwg/libfabric/pull/5742

What I did is address the comments from myself and Robert:

1. Introduced helper function rxr_copy_from_tx() and rxr_copy_to_rx() to wrap rxr_cuda_copy_iov.

2. rename rxr_pkt_send_data_mr_cache to rxr_pkt_send_data_desc() and rxr_prepare_mr_send to rxr_prepare_desc_send() .

### CODE REVIEW

**Comment by rwespetal** [comment]
The CI failure is an issue:
```
15:05:04 # OSU MPI Latency Test v5.6.2
15:05:04 # Size          Latency (us)
15:05:04 0                      18.58
15:05:04 1                      20.49
15:05:05 2                      18.54
15:05:05 4                      18.50
15:05:06 8                      18.54
15:05:06 16                     18.56
15:05:06 32                     18.56
15:05:07 64                     18.58
15:05:07 128                    18.60
15:05:07 256                    18.71
15:05:08 512                    18.88
15:05:08 1024                   19.29
15:05:09 2048                   19.88
15:05:09 4096                   21.06
15:05:09 8192                   23.39
15:05:10 16384                  26.98
15:05:10 32768                  31.40
15:05:10 65536                  38.88
15:05:10 131072                 55.40
15:05:10 262144                131.45
15:05:10 Unable to write to EQ: Missing or unavailable event queue. err: Input/output error (5) prov_errno: Unknown error -5 (-5) prov/efa/src/rxr/rxr.h:947
15:05:10 [ip-172-31-21-6:83024] *** Process received signal ***
15:05:10 [ip-172-31-21-6:83024] Signal: Aborted (6)
15:05:10 [ip-172-31-21-6:83024] Signal code:  (-6)
15:05:10 [ip-172-31-21-6:83024] [ 0] /lib64/libpthread.so.0(+0x117e0)[0x7f66bc9167e0]
15:05:10 [ip-172-31-21-6:83024] [ 1] /lib64/libc.so.6(gsignal+0x110)[0x7f66bc58dae0]
15:05:10 [ip-172-31-21-6:83024] [ 2] /lib64/libc.so.6(abort+0x148)[0x7f66bc58ef88]
15:05:10 [ip-172-31-21-6:83024] [ 3] /home/ec2-user/libfabric/install/lib/libfabric.so.1(+0x9a442)[0x7f66a9625442]
15:05:10 [ip-172-31-21-6:83024] [ 4] /home/ec2-user/libfabric/install/lib/libfabric.so.1(+0x9b718)[0x7f66a9626718]
15:05:10 [ip-172-31-21-6:83024] [ 5] /home/ec2-user/libfabric/install/lib/libfabric.so.1(+0xa1464)[0x7f66a962c464]
15:05:10 [ip-172-31-21-6:83024] [ 6] /home/ec2-user/libfabric/install/lib/libfabric.so.1(+0xa169f)[0x7f66a962c69f]
15:05:10 [ip-172-31-21-6:83024] [ 7] /home/ec2-user/libfabric/install/lib/libfabric.so.1(+0xa1c79)[0x7f66a962cc79]
15:05:10 [ip-172-31-21-6:83024] [ 8] /home/ec2-user/libfabric/install/lib/libfabric.so.1(+0x45c73)[0x7f66a95d0c73]
15:05:10 [ip-172-31-21-6:83024] [ 9] /home/ec2-user/libfabric/install/lib/libfabric.so.1(+0x44eec)[0x7f66a95cfeec]
15:05:10 [ip-172-31-21-6:83024] [10] /home/ec2-user/libfabric/install/lib/libfabric.so.1(+0x451dc)[0x7f66a95d01dc]
15:05:10 [ip-172-31-21-6:83024] [11] /opt/amazon//openmpi/lib64/openmpi/mca_mtl_ofi.so(+0x1e6d)[0x7f66ab9e3e6d]
15:05:10 [ip-172-31-21-6:83024] [12] /opt/amazon//openmpi/lib64/openmpi/mca_pml_cm.so(+0x3be5)[0x7f66b097ebe5]
15:05:10 [ip-172-31-21-6:83024] [13] /opt/amazon//openmpi/lib64/libmpi.so.40(PMPI_Send+0xe6)[0x7f66bd47bb86]
15:05:10 [ip-172-31-21-6:83024] [14] /tmp/osu-micro-benchmarks-ompi/mpi/pt2pt/osu_latency[0x40179b]
15:05:10 [ip-172-31-21-6:83024] [15] /lib64/libc.so.6(__libc_start_main+0xea)[0x7f66bc57b02a]
15:05:10 [ip-172-31-21-6:83024] [16] /tmp/osu-micro-benchmarks-ompi/mpi/pt2pt/osu_latency[0x4018da]
15:05:10 [ip-172-31-21-6:83024] *** End of error message ***
15:05:10 --------------------------------------------------------------------------
15:05:10 Primary job  terminated normally, but 1 process returned
15:05:10 a non-zero exit code. Per user-direction, the job has been aborted.
15:05:10 --------------------------------------------------------------------------
15:05:13 --------------------------------------------------------------------------
15:05:13 mpirun noticed that process rank 1 with PID 83024 on node 172.31.21.6 exited on signal 6 (Aborted).
15:05:13 --------------------------------------------------------------------------
```

File: prov/efa/src/rxr/rxr.h [hunk 1]
```diff
@@ -467,6 +467,43 @@ struct rxr_tx_entry {
...
+	return data_size;
+}
+
+static inline
+size_t rxr_copy_to_rx(void *data, size_t tocopy, struct rxr_rx_entry *rx_entry, size_t offset)
+{
+	size_t data_size;
+#ifdef HAVE_LIBCUDA
+	if (rxr_ep_is_cuda_mr(rx_entry->desc[0]))
+		data_size = ofi_copy_to_cuda_iov(rx_entry->iov, rx_entry->iov_count, offset, data, tocopy);
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr.h]: Can you format this function the same as the tx one?

File: prov/efa/src/rxr/rxr_pkt_type_data.c [hunk 2]
```diff
@@ -147,7 +144,7 @@ ssize_t rxr_pkt_send_data_mr_cache(struct rxr_ep *ep,
 	/* Assign packet header in constructed iov */
 	iov[i].iov_base = rxr_pkt_start(pkt_entry);
 	iov[i].iov_len = sizeof(struct rxr_data_hdr);
-	desc[i] = rxr_ep_mr_local(ep) ? fi_mr_desc(pkt_entry->mr) : NULL;
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_data.c]: This should probably go in a separate PR or another patch in this series where we just remove `rxr_ep_mr_local` completely.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c [hunk 3]
```diff
@@ -271,7 +271,7 @@ void rxr_pkt_handle_readrsp_sent(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_en
 	assert(tx_entry->window >= 0);
 	if (tx_entry->bytes_sent < tx_entry->total_len) {
 		if (efa_mr_cache_enable && rxr_ep_mr_local(ep))
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c]: Doesn't this also need a `rxr_ep_is_cuda_mr` check?

**Comment by wzamazon** [comment]
Thank you! I can reproduce the CI failure, and is working on it.

File: prov/efa/src/rxr/rxr.h (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr.h]: Sure.

File: prov/efa/src/rxr/rxr_pkt_type_data.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_data.c]: Make sense, reverted.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c]: This is the emulated Read path, when GPU Buffer is used, we really should not have go there. I think I will add an assertion here about the mr cannot be a cuda_mr.

**Comment by wzamazon** [comment]
The CI failure earlier is because ```tx_entry->desc``` was not cleaned in ```rxr_tx_entry_init```

File: prov/efa/src/rxr/rxr_pkt_type_misc.c (hunk 3 above)
Comment (rwespetal) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c]: Got it, thanks.

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest
